README.md
# Divvy NodeJS Client

A client for the [Divvy rate limit server](https://github.com/button/divvy).

1. [Requirements](#requirements)
2. [Usage](#usage)
3. [Other Features](#other-features)
4. [License and Copyright](#license-and-copyright)

## Requirements

* NodeJS version 4 or newer.

## Usage

```js
const DivvyClient = require('@button/divvy-client');
const client = new DivvyClient('localhost', 8321);

client.hit({ method: 'GET', path: '/pantry/cookies' }).then((result) => {
    console.log(`Success: ${JSON.stringify(result)}`);
    // { allowed: true, currentCredit: 100, nextResetSeconds: 60 }
}).catch((err) => {
    console.log(`Error: ${err.message}`);    
});
```

## Other Features

### Client Stub

A stub of the interface is exposed as `Client.Stub`, which implements the core methods (connect, close, and hit).

## License and Copyright

Licensed under the MIT license. See `LICENSE.txt` for full terms.

Copyright 2016 Button, Inc.


CHANGELOG.md
# CHANGELOG

## 1.0.1 (2016-10-14)

* Initial public release.


tests/util-test.js
'use strict';

const assert = require('assert');
const util = require('../src/util');

const VALID_OPERATIONS = [
  {},
  { shoes: 'feet'},
  { shoes: 'feet', bangles: 'arm' }
];

const INVALID_OPERATIONS = [
  { hello: new Error() },
  { shoes: { dont: 'lie '} },
  { titanic: 'greatest\nmovie\nevar' },
  { 'i want to be': 'under_the_sea' }
];

describe('src/util', () => {

  describe('#assertValidOperation', () => {

    it('for valid operations', () => {
      for (let oper of VALID_OPERATIONS) {
        util.assertValidOperation(oper);
      }
    });

    it('for invalid operations', () => {
      for (let oper of INVALID_OPERATIONS) {
        const operStr = JSON.stringify(oper);
        /*jshint -W083 */
        assert.throws(() => util.assertValidOperation(oper), /Invalid operation/,
          `Expected operation ${operStr} to be invalid`);
      }
    });

  });

  describe('#operationToString', () => {
    it('for valid operations', () => {
      assert.equal('', util.operationToString({}));
      assert.equal('', util.operationToString(null));
      assert.equal('"hello"="StLouis"', util.operationToString({ hello: 'StLouis' }));
      assert.equal('"a"="1" "b"="2"', util.operationToString({ 'b': 2, 'a': 1 }));
    });

    it('for invalid operations', () => {
      for (let oper of INVALID_OPERATIONS) {
        /*jshint -W083 */
        assert.throws(() => util.operationToString(oper), /Invalid operation/,
          `Expected operation ${oper} to be invalid`);
      }
    });
  });

  describe('#removeNullOrUndefinedKeys', () => {
    it('removes keys', () => {
      assert.deepEqual({
        human: true,
        name: 'Jim',
        age: 75
      }, util.removeNullOrUndefinedKeys({
        bar: null,
        foo: null,
        human: true,
        alien: undefined,
        name: 'Jim',
        age: 75,
        height: undefined
      }));
    });
  });

});


tests/client-test.js
'use strict';

const assert = require('assert');
const net = require('net');
const sinon = require('sinon');

const carrier = require('carrier');

const Errors = require('../src/errors');
const Client = require('../src/client');
const Bluebird = require('bluebird');

describe('src/client', () => {

  describe('general tests', () => {
    // Fake server.
    var server;

    // Test client.
    var client;

    // String the server should received. Set in tests.
    var expectedClientMessage;

    // String the server should send back. Set in tests.
    var mockServerResponse;

    // The server's client connection.
    var clientConnection;

    beforeEach((done) => {
      // Create a server on port 0 (ephemeral / randomly-selected port)
      server = net.createServer((conn) => {
        clientConnection = conn;
        carrier.carry(conn, (line) => {
          assert.equal(expectedClientMessage, line);
          conn.write(mockServerResponse + '\n');
        });
      });

      // Once the server is bound, connect a client to it then finish.
      server.on('listening', () => {
        client = new Client('', server.address().port, {
          autoReconnect: false,
          throttleConnect: false
        });
        client.on('connected', done);
        client.connect();
      });

      server.listen(0);
    });

    it('parses an allowed response', (done) => {
      expectedClientMessage = 'HIT "name"="test" "path"="123"';
      mockServerResponse = 'OK true 1 2';

      client.hit({ name: 'test', path: '123'}).then((response) => {
        assert.deepEqual(response, {
          isAllowed: true,
          currentCredit: 1,
          nextResetSeconds: 2
        });
        done();
      }).catch(done);
    });

    it('sorts HIT operation keys', (done) => {
      expectedClientMessage = 'HIT "name"="test" "path"="123"';
      mockServerResponse = 'OK true 1 2';

      client.hit({ path: '123', name: 'test' }).then((response) => {
        assert.deepEqual(response, {
          isAllowed: true,
          currentCredit: 1,
          nextResetSeconds: 2
        });
        done();
      }).catch(done);
    });

    it('parses a not-allowed response', (done) => {
      expectedClientMessage = 'HIT "name"="test" "path"="123"';
      mockServerResponse = 'OK false 0 0';

      client.hit({ name: 'test', path: '123'}).then((response) => {
        assert.deepEqual(response, {
          isAllowed: false,
          currentCredit: 0,
          nextResetSeconds: 0
        });
        done();
      }).catch(done);
    });

    it('emits an event when disconnected', (done) => {
      client.on('disconnected', () => {
        assert.equal(false, client.connected);
        done();
      });
      setTimeout(() => {
        clientConnection.destroy();
      }, 10);
    });

    it('client automatically connects if needed', (done) => {
      client.close();

      // Create a new client and never explicitly call connect()
      const myClient = new Client('', server.address().port);
      expectedClientMessage = 'HIT "name"="test" "path"="123"';
      mockServerResponse = 'OK true 1 2';

      myClient.hit({ path: '123', name: 'test' }).then((response) => {
        assert.deepEqual(response, {
          isAllowed: true,
          currentCredit: 1,
          nextResetSeconds: 2
        });
        assert(myClient.connected);
        done();
      }).catch(done);

    });
  });

  describe('timeout tests', () => {
    // TODO(mikey): These tests should use a fake clock.

    // Fake server.
    var server;

    // Test client.
    var client;

    beforeEach((done) => {
      // A server that always succeeds, but delays 100ms.
      server = net.createServer((conn) => {
        carrier.carry(conn, () => {
          setTimeout(() => {
            conn.write('OK true 1 2\n');
          }, 50);
        });
      });

      server.on('listening', () => {
        client = new Client('', server.address().port);
        client.on('connected', done);
        client.connect();
      });

      server.listen(0);
    });

    it('processes timeouts correctly', (done) => {
      const promises = [
        client.hit({}, 50).catch((err) => err),
        client.hit({}, 60).catch((err) => err),
        client.hit({}, 25).catch((err) => err),
      ];

      Bluebird.all(promises).then((results) => {
        assert(results[0] instanceof Errors.TimeoutError);

        assert(!(results[1] instanceof Error));
        assert.deepEqual({ isAllowed: true, currentCredit: 1, nextResetSeconds: 2 }, results[1]);

        assert(results[2] instanceof Errors.TimeoutError);
        done();
      }).catch(done);

    });
  });

  describe('autoreconnect tests', () => {
    // Fake server.
    var server;
    var serverPort;

    var clock;

    beforeEach((done) => {

      // A server that tells us when a client has connected.
      server = net.createServer(function(conn) {
        this.emit('client-connected', conn);
      });

      server.on('listening', () => {
        serverPort = server.address().port;
        done();
      });

      server.listen(0);

      clock = sinon.useFakeTimers();
    });

    afterEach(() => {
      clock.restore();
    });

    it('reconnects automatically', (done) => {
      var reconnectCount = 0;

      const client = new Client('', serverPort, {
        autoReconnect: true,
        throttleConnect: false
      });

      // Client will keep reconnecting.
      server.on('client-connected', (conn) => {
        reconnectCount++;
        if (reconnectCount === 10) {
          done();
          return;
        }
        conn.destroy();
      });

      client.connect();
    });

    it('throttles reconnect', (done) => {
      const client = new Client('', serverPort, {
        autoReconnect: true,
        throttleConnect: true
      });

      // Client will keep reconnecting.
      server.on('client-connected', (conn) => {
        conn.destroy();
      });

      client.once('disconnected', () => {
        assert(!client.connected);
        clock.tick(1000);
        client.once('connected', () => {
          assert(client.connected);
          done();
        });
      });

      client.connect();
    });

  });

  describe('maxPendingRequests tests', () => {
    // Fake server.
    var server;
    var serverPort;

    beforeEach((done) => {
      server = net.createServer(function(conn) {
        this.emit('client-connected', conn);
        carrier.carry(conn, () => {
          setTimeout(() => {
            conn.write('OK true 1 2\n');
          }, 100);
        });
      });

      server.on('listening', () => {
        serverPort = server.address().port;
        done();
      });

      server.listen(0);
    });

    it('rejects after 3 pending requests', (done) => {
      const client = new Client('', serverPort, {
        autoReconnect: true,
        maxPendingRequests: 3
      });

      const promises = [
        client.hit({}).catch((err) => err),
        client.hit({}).catch((err) => err),
        client.hit({}).catch((err) => err),
        client.hit({}).catch((err) => err),
      ];

      Bluebird.all(promises).then((results) => {
        assert(!(results[0] instanceof Error));
        assert(!(results[1] instanceof Error));
        assert(!(results[2] instanceof Error));
        assert(results[3] instanceof Errors.BacklogError);
        done();
      }).catch(done);
    });

    // Test is flaky on travis :-(
    xit('rejects in-flight requests when disconnected', (done) => {
      const client = new Client('', serverPort, {
        autoReconnect: false,
        maxPendingRequests: 3
      });

      server.once('client-connected', (conn) => {
        conn.destroy();
      });

      client.once('connected', () => {
        const promises = [
          client.hit({}).catch((err) => err),
          client.hit({}).catch((err) => err)
        ];

        Bluebird.all(promises).then((results) => {
          assert(results[0] instanceof Errors.DisconnectedError);
          assert(results[1] instanceof Errors.DisconnectedError);
          done();
        }).catch(done);
      });

      client.connect();

    });

  });

  describe('Client.Stub', function() {

    beforeEach(function() {
      this.client = new Client.Stub();
    });

    it('is a no-op hit', function(done) {
      this.client.hit({}).then(quota => {
        assert.equal(quota.isAllowed, true);
        assert.equal(quota.currentCredit, 0);
        assert.equal(quota.nextResetSeconds, 0);
        done();
      });
    });

    it('does not error when calling connect', function() {
      this.client.connect();
    });

    it('does not error when calling close', function() {
      this.client.close();
    });

  });

});


tests/jshint-test.js
'use strict';

require('mocha-jshint')({ pretty: true });

src/util.js
'use strict';

// Whitespace and quote characters are not allowed.
const KEY_VALUE_RE = /^[^\s"]+$/;

module.exports = {

  assertValidOperation: (operation) => {
    for (let key in operation) {
      module.exports.assertValidKeyValue(key, operation[key]);
    }
  },

  assertValidKeyValue: (key, value) => {
    if (!KEY_VALUE_RE.test(key)) {
      throw new Error(`Invalid operation key: "${key}"`);
    }

    if (typeof value === 'number') {
      return;
    } else if (typeof value !== 'string') {
      throw new Error(`Invalid operation value for key {$key}: not string or number: ${value}`);
    }

    if (!KEY_VALUE_RE.test(value)) {
      throw new Error(`Invalid operation value for key "${key}": "${value}"`);
    }
  },

  operationToString: (operation) => {
    operation = operation || {};
    module.exports.assertValidOperation(operation);
    const pairs = Object.keys(operation).sort().map(k => `"${k}"="${operation[k]}"`);
    return pairs.join(' ');
  },

  removeNullOrUndefinedKeys: operation => {
    const updated = {};

    Object.keys(operation).forEach(key => {
      if (operation[key] !== null && operation[key] !== undefined) {
        updated[key] = operation[key];
      }
    });

    return updated;
  }

};


src/errors.js
'use strict';

class DivvyClientError extends Error {
  constructor(message) {
    super(message);
    this.name = this.constructor.name;
    this.message = message;
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = (new Error(message)).stack;
    }
  }
}

/** Message timed out. */
class TimeoutError extends DivvyClientError {
}

/** Tried to send message while disconnected. */
class DisconnectedError extends DivvyClientError {
}

/** Too many requests in flight. */
class BacklogError extends DivvyClientError {
}

/** Got a bad response from the server. */
class BadResponseError extends DivvyClientError {
}

module.exports = {
  DivvyClientError: DivvyClientError,
  TimeoutError: TimeoutError,
  DisconnectedError: DisconnectedError,
  BacklogError: BacklogError,
  BadResponseError: BadResponseError
};

src/client.js
'use strict';

const carrier = require('carrier');
const net = require('net');
const Errors = require('./errors');
const util = require('./util');
const EventEmitter = require('events').EventEmitter;

function defaultIfUndefined(val, defaultVal) {
  return val !== undefined ? val : defaultVal;
}

/**
 * Basic Divvy protocol client.
 *
 * Commands supported: The only command currently supported is `hit()`.
 *
 * Connection management: The client can be explicitly connected by
 * calling `connect()`. The connection will be kept open as long as possible.
 * If a command is called while the client is disconnected, the command
 * will be enqueued and the client will be connected.
 *
 * Timeouts:
 */
class Client extends EventEmitter {

  /**
   * Constructor.
   *
   * @param  {string} host    server hostname
   * @param  {number} port    server port number
   * @param  {boolean} options.autoReconnect  whether to automatically reconnect when the
   *     server connection is unexpectedly closed (default true)
   * @param  {number} options.defaultCommandTimeoutMillis  default timeout for commands, or
   *     `null` to not use a timouer (default 1000)
   * @param  {number} options.maxPendingRequests  maximum number of outgoing requests
   *     that are allowed to be in flight at any one time. Requests made over this
   *     limit will be dropped. (default 100)
   * @param  {number} options.maxReconnectAttempts  when `options.autoReconnect` is true,
   *     maximum number of consecutive `connect()` attempts the client will automatically
   *     make before giving up
   * @param  {boolean} options.throttleConnect  whether to throttle calls to `connect()`
   *     such that they will not be attempted more than once every
   *     `options.throttleConnectTimeoutMillis` (default true)
   * @param  {boolean} options.throttleConnectTimeoutMillis  when `throttleConnect` is
   *     `true`, enforce a delay of this many milliseconds before connecting again.
   */
  constructor(host, port, options) {
    super();

    options = options || {};

    this.host = host || 'localhost';
    this.port = port || 8321;

    this.clientSocket = null;
    this.connected = false;

    // Requests that have not yet been written to the socket. Typically
    // empty unless disconnected.
    this.requestQueue = [];

    // Requests that have been written to the socket and are awaiting a
    // response.
    this.responseQueue = [];

    this.autoReconnect = defaultIfUndefined(options.autoReconnect, true);
    this.maxReconnectAttempts = defaultIfUndefined(options.maxReconnectAttempts, 5);
    this.throttleConnect = defaultIfUndefined(options.throttleConnect, true);
    this.throttleConnectTimeoutMillis = defaultIfUndefined(
      options.throttleConnectTimeoutMillis, 1000);
    this.maxPendingRequests = defaultIfUndefined(options.maxPendingRequests, 100);
    this.defaultCommandTimeoutMillis = defaultIfUndefined(
      options.defaultCommandTimeoutMillis, 1000);

    // Timeout handle, set when the client is unexpectedly closed.
    this.connectTimeoutHandle = null;
    this.lastDisconnectDate = null;
    this.reconnectAttempts = 0;
    this.onDisconnectedListener = this._onUnexpectedDisconnect.bind(this);
  }

  /** Schedules connection to the server; no-op if already connected. */
  connect() {
    if (this.clientSocket || this.connectTimeoutHandle) {
      // Already connected or connecting. Ignore.
      return;
    }

    if (!this.throttleConnect) {
      this._doConnect();
      return;
    }

    var delay = 0;
    if (this.lastDisconnectDate) {
      let age = Math.max(0, new Date() - this.lastDisconnectDate);
      delay = Math.max(0, this.throttleConnectTimeoutMillis - age);
      this.lastDisconnectDate = null;
    }

    if (delay <= 0) {
      this._doConnect();
    } else {
      this.connectTimeoutHandle = setTimeout(() => this._doConnect(), delay);
    }
  }

  /** Manually close the connection. This will not trigger a reconnect. */
  close() {
    if (this.connectTimeoutHandle) {
      clearTimeout(this.connectTimeoutHandle);
      this.connectTimeoutHandle = null;
    }

    if (!this.clientSocket) {
      return;
    }

    this.clientSocket.removeListener('close', this.onDisconnectedListener);
    this.clientSocket.destroy();
    this._doDisconnect();
  }

  /**
   * Perform a "hit" command against the given operation.
   * Upon success, the promise is resolve with an object containing
   * fields `isAllowed` (boolean), `currentCredit`, and
   * `nextResetSeconds`.
   *
   * @param  {object} operation the operation object, consisting of string key-value pairs
   *                            (optional, default: `{}`)
   * @param  {number} timeout   the timeout in millis; if undefined, uses instance
   *                            default timeout.
   */
  hit(operation, timeout) {
    operation = util.removeNullOrUndefinedKeys(operation || {});

    if (timeout === undefined) {
      timeout = this.defaultCommandTimeoutMillis;
    }

    const numPending = this._numPendingRequests();
    if (numPending >= this.maxPendingRequests) {
      return Promise.reject(new Errors.BacklogError(`Too many pending requests (${numPending})`));
    }

    const operStr = util.operationToString(operation);

    var message;
    if (operStr) {
      message = `HIT ${operStr}\n`;
    } else {
      message = 'HIT\n';
    }

    const pendingRequest = this._enqueueMessage(message, timeout);
    return pendingRequest.promise;
  }

  /** Returns total number of outstanding requests. */
  _numPendingRequests() {
    // When connected, the requestQueue will typically be near-empty
    // (commands are flushed quickly) and the responseQueue will match
    // the arrival rate.
    //
    // When disconnected, the requestQueue will fill up and the responseQueue
    // will be empty (no writes to socket until connected).
    return this.requestQueue.length + this.responseQueue.length;
  }

  _doConnect() {
    this.connectTimeoutHandle = null;
    this.clientSocket = new net.Socket();

    this.clientSocket.connect(this.port, this.host, this._onConnected.bind(this));
    this.clientSocket.on('close', this.onDisconnectedListener);
    this.clientSocket.on('error', (err) => {
      this.emit('error', err);
    });
  }

  _onConnected() {
    this.connected = true;
    this.reconnectAttempts = 0;

    carrier.carry(this.clientSocket, (line) => {
      this._receivedLine(line);
    });
    this.emit('connected');

    // One or more requests could have been enqueued while waiting to connect.
    this._flushPending();
  }

  /** Do cleanup needed with every disconnect. */
  _doDisconnect() {
    this.connected = false;
    this.clientSocket = null;
    this._rejectAllPending();
    this.emit('disconnected');
  }

  _onUnexpectedDisconnect() {
    this.lastDisconnectDate = new Date();
    this._doDisconnect();
    if (this.autoReconnect && this.reconnectAttempts < this.maxReconnectAttempts) {
      this.reconnectAttempts++;
      this.connect();
    }
  }

  /** Immediately rejects all requests, whether or not they have been sent. */
  _rejectAllPending() {
    while (this.requestQueue.length) {
      let elem = this.requestQueue.shift();
      if (elem.isRejectedOrResolved) {
        continue;
      }
      elem.reject(new Errors.DisconnectedError('Connection closed.'));
    }

    while (this.responseQueue.length) {
      let elem = this.responseQueue.shift();
      if (elem.isRejectedOrResolved) {
        continue;
      }
      elem.reject(new Errors.DisconnectedError('Connection closed.'));
    }
  }

  /**
   * Parses a protocol response.
   * @param  {[type]} line [description]
   * @return {[type]}      [description]
   */
  _parseLine(line) {
    const tokens = line.split(' ');

    if (tokens.length !== 4 || tokens[0] !== 'OK') {
      return new Errors.BadResponseError(line);
    }

    const isAllowed = tokens[1] === 'true';
    const currentCredit = parseInt(tokens[2], 10);
    const nextResetSeconds = parseInt(tokens[3], 10);

    return {
      isAllowed: isAllowed,
      currentCredit: currentCredit,
      nextResetSeconds: nextResetSeconds
    };
  }

  _receivedLine(line) {
    const currentRequest = this.responseQueue.shift();

    if (!currentRequest) {
      throw new Errors.BadResponseError('Bug: Received an unexpected response.');
    }

    // If we already rejected the request in flight,
    if (currentRequest.isRejectedOrResolved) {
      return;
    }

    const response = this._parseLine(line);
    if (response instanceof Error) {
      currentRequest.reject(response);
    } else {
      currentRequest.resolve(response);
    }
  }

  _enqueueMessage(message, timeout) {
    const pendingRequest = this._newPendingRequest(message, timeout);
    this.requestQueue.push(pendingRequest);
    if (!this.clientSocket) {
      this.connect();
    } else if (this.connected) {
      this._flushPending();
    }
    return pendingRequest;
  }

  _flushPending() {
    while (this.requestQueue.length) {
      let pendingRequest = this.requestQueue.shift();
      if (pendingRequest.isRejectedOrResolved) {
        // Request timed out, don't even both sending.
        continue;
      }
      this.clientSocket.write(pendingRequest.message);
      this.responseQueue.push(pendingRequest);
    }
  }

  _newPendingRequest(message, timeout) {
    const pendingRequest = {
      message: message,
      isRejectedOrResolved: false
    };

    pendingRequest.promise = new Promise((resolve, reject) => {
      pendingRequest.resolve = function(obj) {
        if (pendingRequest.isRejectedOrResolved) {
          return;
        }
        if (pendingRequest.timeout) {
          clearTimeout(pendingRequest.timeout);
        }
        pendingRequest.isRejectedOrResolved = true;
        resolve(obj);
      };

      pendingRequest.reject = function(err) {
        if (pendingRequest.isRejectedOrResolved) {
          return;
        }
        if (pendingRequest.timeout) {
          clearTimeout(pendingRequest.timeout);
        }
        pendingRequest.isRejectedOrResolved = true;
        reject(err);
      };

      if (timeout !== undefined && timeout !== null) {
        pendingRequest.timeout = setTimeout(function() {
          pendingRequest.reject(new Errors.TimeoutError('Timeout'));
        }, timeout);
      }
    });

    return pendingRequest;
  }

}

/**
 * Stub of the public interface.
 */
Client.Stub = class DivvyClientStub extends Client {
  connect() {
  }

  close() {
  }

  hit() {
    return Promise.resolve({
      isAllowed: true,
      currentCredit: 0,
      nextResetSeconds: 0
    });
  }
};

/** Expose error hierarchy. */
Client.Error = Errors;

module.exports = Client;
